Local patch to allow us to build in unprivileged containers where we
can't make device nodes. The GRUB_PERMIT_NON_BLOCK_DEVICES environment
variable (set in our makevm script) changes behaviour in grub-install.

diff --git a/grub-core/osdep/linux/getroot.c b/grub-core/osdep/linux/getroot.c
index 001b818..e950edd 100644
--- a/grub-core/osdep/linux/getroot.c
+++ b/grub-core/osdep/linux/getroot.c
@@ -725,10 +725,13 @@ grub_util_part_to_disk (const char *os_dev, struct stat *st,
 {
   char *path;
 
-  if (! S_ISBLK (st->st_mode))
+  if (! getenv("GRUB_PERMIT_NON_BLOCK_DEVICES"))
     {
-      *is_part = 0;
-      return xstrdup (os_dev);
+      if (! S_ISBLK (st->st_mode))
+        {
+          *is_part = 0;
+          return xstrdup (os_dev);
+        }
     }
 
   path = xmalloc (PATH_MAX);
diff --git a/grub-core/osdep/linux/hostdisk.c b/grub-core/osdep/linux/hostdisk.c
index 7e24ae6..40f4e68 100644
--- a/grub-core/osdep/linux/hostdisk.c
+++ b/grub-core/osdep/linux/hostdisk.c
@@ -99,10 +99,41 @@ sysfs_partition_path (const char *dev, const char *entry)
 {
   struct stat st;
 
-  if (stat (dev, &st) == 0 && S_ISBLK (st.st_mode))
-    return xasprintf ("/sys/dev/block/%u:%u/%s",
-		      major (st.st_rdev), minor (st.st_rdev), entry);
+  // grub_util_info ("sysfs_partition_path: dev %s entry %s", dev, entry);
 
+  if (! getenv("GRUB_PERMIT_NON_BLOCK_DEVICES"))
+    {
+      if (stat (dev, &st) != 0) {
+        // grub_util_info ("sysfs_partition_path: stat failed");
+	return NULL;
+      }
+
+      if (! S_ISBLK (st.st_mode)) {
+        // grub_util_info ("sysfs_partition_path: not a block device");
+	return NULL;
+      }
+
+      // grub_util_info ("sysfs_partition_path: returning stuff");
+      return xasprintf ("/sys/dev/block/%u:%u/%s",
+			major (st.st_rdev), minor (st.st_rdev), entry);
+    }
+  else
+    {
+      if (strncmp("/dev/sda", dev, 8) == 0)
+        {
+          int major = 8; // Hard-coded for /dev/sda for now!
+          int minor;
+
+          char *p = (char *)dev + 8;
+          minor = strtoul(p, NULL, 10);
+
+          grub_util_info ("sysfs_partition_path: returning faked data for /dev/sdaX");
+          return xasprintf ("/sys/dev/block/%d:%d/%s",
+			major, minor, entry);
+        }
+    }
+
+  grub_util_info ("sysfs_partition_path: not /dev/sdaX...");
   return NULL;
 }
 
@@ -115,12 +146,16 @@ sysfs_partition_start (const char *dev, grub_disk_addr_t *start)
   int ret = 0;
 
   path = sysfs_partition_path (dev, "start");
-  if (!path)
+  if (!path) {
+    // grub_util_info ("sysfs_partition_start: no path");
     return 0;
+  }
 
   fp = grub_util_fopen (path, "r");
-  if (!fp)
+  if (!fp) {
+    // grub_util_info ("sysfs_partition_start: can't open %s", path);
     goto out;
+  }
 
   if (fscanf (fp, "%llu", &val) == 1)
     {
@@ -128,6 +163,8 @@ sysfs_partition_start (const char *dev, grub_disk_addr_t *start)
       ret = 1;
     }
 
+  // grub_util_info ("sysfs_partition_start: *start %lu", *start);
+
 out:
   free (path);
   if (fp)
@@ -143,8 +180,10 @@ grub_util_find_partition_start_os (const char *dev)
   grub_util_fd_t fd;
   struct hd_geometry hdg;
 
-  if (sysfs_partition_start (dev, &start))
+  if (sysfs_partition_start (dev, &start)) {
+    // grub_util_info ("grub_util_find_partition_start_os: start %lu", start);
     return start;
+  }
 
   fd = open (dev, O_RDONLY);
   if (fd == -1)
